File: src/model/spirit.js
'use strict';
// Spirit — a rune's identity: a frozen random real-ID + a human-readable name.
// See ARCHITECTURE.md §2.1.

const crypto = require('crypto');

// Mint a real ID: a prefix + random hex. Stable, unique, never reused.
function mintId(prefix = 'rune') {
  return `${prefix}_${crypto.randomBytes(6).toString('hex')}`;
}

// Create a spirit. If no id is given, one is minted. `name` is the human handle
// (doubles as a tag), must be unique within a mantle (enforced by the mantle).
function createSpirit(name, id) {
  if (!name || typeof name !== 'string') {
    throw new Error('spirit requires a name');
  }
  return { id: id || mintId(), name };
}

module.exports = { mintId, createSpirit };

File: core/src/model/spirit.c
/* spirit.c — a rune/mantle identity: a frozen real-ID + an editable name.
 * SPEC §3.1. */
#include "vc_internal.h"

cJSON *vc_spirit_new(const char *prefix, const char *name) {
  char id[64];
  vc_mint_id(prefix ? prefix : "rune", id, sizeof id);
  cJSON *s = cJSON_CreateObject();
  cJSON_AddStringToObject(s, "id", id);
  cJSON_AddStringToObject(s, "name", name ? name : "");
  return s;
}
